.github/workflows/build.yml

```diff
@@ -0,0 +1,92 @@
+name: Branch Build
+
+on:
+  push:
+    branches:
+    - master
+    - release-*
+
+env:
+  CONTAINER_REGISTRY: us-docker.pkg.dev/spinnaker-community/docker
+
+jobs:
+  lint:
+    # Only run this on repositories in the 'spinnaker' org, not on forks.
+    if: startsWith(github.repository, 'spinnaker/')
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v2
+    - name: Ensure code formatting and style is consistent
+      uses: golangci/golangci-lint-action@v2
+      with:
+        version: v1.32
+
+  branch-build:
+    # Only run this on repositories in the 'spinnaker' org, not on forks.
+    if: startsWith(github.repository, 'spinnaker/')
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Extract repository name
+        id: extract_repo_name
+        run: echo ::set-output name=REPO::${GITHUB_REPOSITORY##*/}
+      - name: Get date
+        id: get_date
+        run: echo ::set-output name=DATETIME::$(date --utc +'%Y%m%d%H%M')
+
+      - uses: actions/setup-go@v2
+        with:
+          go-version: 1.15
+      - uses: actions/cache@v2
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: Fetch dependencies
+        run: go get -d -v
+      - name: Test
+        run: go test -v ./...
+      - name: Build binaries
+        env:
+          LDFLAGS: "-X github.com/spinnaker/spin/version.Version=${{ github.ref_name }}-${{ github.sha }}-${{ steps.get_date.outputs.DATETIME }}"
+        run: |
+          GOARCH=amd64 GOOS=darwin go build -ldflags "${LDFLAGS}" -o dist/${{ github.ref_name }}/${{ github.sha }}-${{ steps.get_date.outputs.DATETIME }}/darwin/amd64/spin .
+          GOARCH=amd64 GOOS=linux go build -ldflags "${LDFLAGS}" -o dist/${{ github.ref_name }}/${{ github.sha }}-${{ steps.get_date.outputs.DATETIME }}/linux/amd64/spin .
+          GOARCH=amd64 GOOS=windows go build -ldflags "${LDFLAGS}" -o dist/${{ github.ref_name }}/${{ github.sha }}-${{ steps.get_date.outputs.DATETIME }}/windows/amd64/spin.exe .
+          dist/${{ github.ref_name }}/${{ github.sha }}-${{ steps.get_date.outputs.DATETIME }}/linux/amd64/spin --version
+      - name: Login to Google Cloud
+        uses: 'google-github-actions/auth@v0'
+        # use service account flow defined at: https://github.com/google-github-actions/upload-cloud-storage#authenticating-via-service-account-key-json
+        with:
+          credentials_json: '${{ secrets.GAR_JSON_KEY }}'
+      - name: Upload spin CLI binaries to GCS
```

Comment:
It is questionable about whether we want to publish binaries and container images on every merge `master` and `release-*` branch. It does create a lot of artifacts that one might want to sort through.

Currently in the [GCS Bucket](https://console.cloud.google.com/storage/browser/spinnaker-artifacts/spin?cloudshell=false&project=spinnaker-community) there is a "top level" folder convention of:
1. `spin/<tag>/...`, and
2. `spin/<tag>-<date>/...` - << perhaps this was done on merge master and other cadences? Can't tell because Jenkins is dead.

It's possible to sort the bucket to get to final releases but by default `1.17` will be at the end of the list of 800+ items shown.

The naming convention in this PR puts `master` and `release-*` commits into subfolder's.
This makes it easier to browse the bucket and filter for an artifact "in between" official releases. One can also search by git sha, etc.
It would be even easier to casually browse through if older `spin/<tag>-<date>/` versions were deleted or moved to their own "older" folder. 

1. `spin/<tag>/...` for releases,
2. `spin/master/<git sha>-<date time>/...` 
3. `spin/release-1.26.x/<git sha>-<date time>/...` 
3. `spin/release-1.27.x/<git sha>-<date time>/...`